src/dataset.cpp

```diff
@@ -337,6 +338,24 @@ void multiply(const gsl::index size, ptr<double> RESTRICT v1,
     v1[i] *= v2[i];
   }
 }
+
+void multiply(VariableSlice &lhs, const ConstVariableSlice &rhs) { lhs *= rhs; }
+
+void divide(const gsl::index size, ptr<double> RESTRICT v1,
+            ptr<double> RESTRICT e1, ptr<const double> RESTRICT v2,
+            ptr<const double> RESTRICT e2) {
+  for (gsl::index i = 0; i < size; ++i) {
+    e1[i] = e1[i] * (v2[i] * v2[i]) + e2[i] * (v1[i] * v1[i]);
```

Comment:
I'll repeat the point I made on slack here for the record.

I've thought a bit more about propagation of errors and i don't think we can just go around multiplying variances (and their units) together.

If you look at the formulas for propagation of uncertainty (e.g. wikipedia https://en.wikipedia.org/wiki/Propagation_of_uncertainty), the units of the variance are always `f^2`, what is inside the square brackets is dimensionless (we assume `sig_AB` is zero).

Say we have dataset `A` contains a value in `m`, and the same for `B`. if you do `A * B` you'll get a variance with units `m^4` which is fine, it's `(A*B)^2`

But if you then divide by B, the dimensions of the variance of `f = A*B / B = A` is just `m^2`. So you add uncertainty by doing `A*B/B` because each operation adds errors in quadrature, but the units of a variance is always the square of the resulting operation, it doesn't depend on anything else.

So i don't think we can just go around multiplying variances together inside datasets, we have to only operate on the `Value`, propagate the errors in quadrature, and give them the units of `Value^2`

Reply:
I am not sure what the issue is here? The linked wikipedia article gives as equation for the variance of a division `A/B`:

```
var(A/B) = (A/B)^2 [ var(A)/A^2 + var(B)/B^2 ]
```

If A is `m^2` and and B is `m` we get:

```
var(A/B) = m^4/m^2 [ m^4/m^2 + m^2/m] = m^2
``` 
as required. Thus, we *can* just multiply things together and there is no need for special handling of the unit.

Reply:
but when trying with `counts`, `counts^4` is not part of the allowed units

Reply:
Dividing `counts/counts` should give:

```
var(A/B) = m^2/m^2 [ m^2/m^2 + m^2/m^2] = dimensionless
```

For certain order of execution of operations you may have an intermediate `m^4`, is it sufficient to just reorder the operations?

Reply:
Yes I think the problem is exactly the intermediate unit. I'll take a look into this if I get time today or tomorrow.

Reply:
I would not bother with this now. As I mentioned previously the operations on variables groups will be refactored (see item 4 in https://github.com/mantidproject/dataset/issues/58), so we should not waste time with this now.

Reply:
Ok. Maybe one easy adaption/diagnostic would be to allow u^4 to support the intermediate operations and see if that solves the issues. https://github.com/mantidproject/dataset/blob/master/src/unit.h#L150

Reply:
No that would make things worse, because you can then multiply counts with counts, which does not make sense.

Reply:
i agree that re-ordering the operation should work

Reply:
Agreed. I wasn't suggesting actually leaving that change in place.